OSDOCS-6534: adding CSI snapshotting (#62557)
openshift/openshift-docs: modules/microshift-storage-classes.adoc

// Module included in the following assemblies:
//
// microshift/volume-snapshots-microshift.adoc

:_content-type: CONCEPT
[id="microshift-storage-classes_{context}"]

> == Storage classes

Reviewer: In terms of reuse, you want to start this with an H1 and then use the `[leveloffset=+2]` in the assembly to get the correct information architecture.

Reviewer 2: +1

Storage classes provide the workload layer interface for selecting a device class. The following storage class parameters are supported in {product-title}:

* `csi.storage.k8s.io/fstype`, which selects the filesystem types. `xfs` or `ext4` filesystem type sare supported
* `topolvm.io/device-class` is the name of the device class. If a device class is not provided, the default device class is assumed.

Multiple storage classes can refer to the same device class. You can provide varying sets of parameters for the same backing device class, such as `xfs` and `ext4` variants.

> .{product-title} defines the following as the default storage class:

Reviewer 3: @copejon, does the user need to provide an additional StorageClass for snapshotting? If so, we need to provide an example of what that looks like.

Reviewer 3: This is a kube API resource. It exists in the database, but not in a separate configuration file. We want to document the defaults in this format because it's the same format a user would need to use to add their own storage class.

> 

Reviewer: It might help to orient customers by labeling the example with a file name, such as `.Example persistent volume claim (PVC)` or `.Example storage class resource`

[source, yaml]
----
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    storageclass.kubernetes.io/is-default-class: "true" <1>
  name: topolvm-provisioner
parameters:
  "csi.storage.k8s.io/fstype": "xfs" <2>
provisioner: topolvm.io <3>
reclaimPolicy: Delete
volumeBindingMode: WaitForFirstConsumer <4>
allowVolumeExpansion: false <5>
----
[.small]

> <1> This storage class is the cluster's default. If a PVC does not specify a storage class, this class is assumed. Delete “true” if you are not using the {product-title} default storage class. You must set this if you are defining a storage class.

Reviewer 3: I think we want them to delete the annotation entirely, not just the `"true"` part.

```suggestion
<1> This storage class is the cluster's default. If a PVC does not specify a storage class, this class is assumed. Delete the `storage class.kubernetes.io/is-default-class` annotation if you are not using the {product-title} default storage class. You must set this if you are defining a storage class.
```

Author: Used Jon's suggestion that follows.

> <1> An example of {product-title}'s default storage class. If a PVC does not specify a storage class, this class is assumed. There can only be one default storage class in a cluster. Having no value assigned to this annotation is also supported.

Reviewer 2: I'd reword to avoid "Microshift's". I think you could probably just say "An example of the default storage class."

> <1> This storage class is the cluster's default. If a PVC does not specify a storage class, this class is assumed. Delete the `storage class.kubernetes.io/is-default-class` annotation if you are not using the {product-title} default storage class. You must set this if you are defining a storage class.

Reviewer 4: ```suggestion
<1> This is an example of {product-title}'s default storage class. If a PVC does not specify a storage class, this class is assumed.
```

The default annotation is purely optional.  Only 0 or 1 storage classes in the cluster can be designated as default.  

I didn't see reference to storage class docs, but I think it'd be good to point users there for more details.

https://docs.openshift.com/container-platform/4.13/storage/dynamic-provisioning.html#defining-storage-classes_dynamic-provisioning

> <2> Which filesystem to provision on the volume. Options are "xfs" and "ext4".

Reviewer: ```suggestion
<2> Specifies which filesystem to provision on the volume. Options are "xfs" and "ext4".
```

> <2> Which filesystem to provision on the volume. Options are "xfs" and "ext4".

Reviewer 2: Mentioned above, file system should likely be two words (per ISG), unless there's a reason it shouldn't be in this context

<3> Identifies which provisioner should manage this class.

> <4> Whether to provision the volume before a client pod is present or immediately. Options are `WaitForFirstConsumer` and `Immediate`. `WaitForFirstConsumer` is recommended to ensure that storage is only provisioned for pods that can be scheduled.

Reviewer: ```suggestion
<4> Specifies whether to provision the volume before a client pod is present or immediately. Options are `WaitForFirstConsumer` and `Immediate`. `WaitForFirstConsumer` is recommended to ensure that storage is only provisioned for pods that can be scheduled.
```

> <5> Specifies whether the storage provider supports volume expansion. The {product-title} CSI plugin does not support volume expansion, so this field has no effect.

Reviewer 5: Checked on MicroShift cluster, it does support volume expansion with default storage class, but if storage-class is configured with `allowVolumeExpansion: false`, volume expansion is not allowed.

Author: @copejon Can you answer this one?

Reviewer 4: @radeore is right - LVMS does support volume expansion.

Author: I emptied this field and updated the note. PTAL, thx!
